nmsg/statsmod_plugin.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2024 by Domaintools, LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef NMSG_STATSMOD_PLUGIN_H
+#define NMSG_STATSMOD_PLUGIN_H
+
+/*! \file nmsg/statsmod_plugin.h
+ * \brief Implementing statistics export modules.
+ *
+ * This file defines the interface that developers of statistics export modules
+ * must implement. For the interface for loading and calling statistics modules,
+ * see nmsg/fltmod.h.
+ *
+ * Statistics modules are dynamically loaded shared objects that must export a
+ * symbol called <tt>nmsg_statsmod_plugin_export</tt>. This is a structure of
+ * type #nmsg_statsmod_plugin and is the sole entry point into the module.
+ *
+ * The first field of the nmsg_statsmod_plugin structure is the version of the
+ * API/ABI between libnmsg and the statistics module. Module developers should
+ * assign this field the value #NMSG_STATSMOD_VERSION, or they can add
+ * <tt>NMSG_STATSMOD_REQUIRED_INIT,</tt> to the initializer, which is a
+ * convenience macro that initializes required fields.
+ *
+ * A statistics module (statsmod) needs to provide four functions:
+ *  1. An initialization function function <tt>module_init</tt>,
+ *  2. A finalization function <tt>module_fini<tt>,
+ *  3. A function to add an nmsg_io to instrumentation <tt>io_add</tt>, and
+ *  4. A function to remove an nmsg_io from instrumentation <tt>io_remove</tt>.
+ *
+ * A statistics module is expected to call `nmsg_io_get_stats()` on any enrolled
+ * nmsg_io object and expose these stats appropriately.
+ */
+
+#include <nmsg.h>
+
+/** Version number of the nmsg statsmod ABI. */
+#define NMSG_STATSMOD_VERSION	1
+
+/**
+ * Initialize the stats module.
+ *
+ * Data with module-defined meaning may be passed in via the 'param' and
+ * 'len_param' parameters. This can be used to, for example, configure
+ * module-specific statistics export parameters.
+ *
+ * \param[in] io
+ *	nmsg_io_t object to instrument.
+ *
+ * \param[in] param
+ *	Module-defined data needed for the initialization of the module.
+ *
+ * \param[in] len_param
+ *	Length of 'param'.
+ *
+ * \param[out] mod_data
+ *	Module-defined, module-wide state, passed to other module functions
+ *	that take a 'mod_data' parameter.
+ *
+ * \return #nmsg_res_success
+ *	If the module was successfully initialized.
+ * \return
+ *	Any other result to indicate a fatal error.
+ */
+typedef nmsg_res
+(*nmsg_statsmod_module_init_fp)(const void *param,
+				size_t len_param,
```

Comment:
minor... `const`

Reply:
The filter module API has the `len_param` parameter as `const`, but I believe this was an error -- `len_param` is passed by value, so the `const` only prevents the called function from modifying _its own copy_ of the value, and has no effect on (or benefit to) the caller.